### ISSUE TITLE
Add Drone Agent build filtering

### ISSUE THREAD

**Issue by gtaylor** [drone/drone#2171]
Particularly in cloud environments with instance-based IAM systems, it's important to make sure that builds run on the instance with the most appropriate IAM profiles relative to the build. For example:

* If repo1 requires pushing artifacts to an S3 or GS bucket, it'll need to run on an instance that has the correct IAM profile.
* If repo2 does not require elevated privs, it should not run on an instance with IAM privs that it does not need.

While this could be managed with .drone.yaml labels, doing this at the agent level for critical/sensitive builds would lessen the likelihood of external change. For example:

* I set some labels in my .drone.yaml, pointing at a particular subset of my cluster that is only for this build.
* Someone sneaks a label change into one of my hundreds of other repos, thus giving the malicious party an open avenue for messing with whatever privs my IAM profile grants.

I won't be very prescriptive about implementation details, but here are the things that would be great to be able to do:

* Glob org and/or repo names: `myorg/some*`, `my*/someprefix-*`. This can be potentially dangerous in environments without strict control over repo creation, so we may consider including a note warning as such.
* Set multiple patterns to match: `myorg/repo1,myorg/repo2`

**Comment by ozbillwang** [comment]
In our case, I need Drone agents to report itself with a name (label or tag), for example, `agent-dev`, `agent-stag`, agent-prod`

So if I need it to be built for dev environment drone server will send the job to these `drone-dev` agents.

Pipeline file should be something like this

```diff
pipeline:
  backend:
    image: golang
    commands:
      - go build
      - go test
+   agent:
+    - drone-dev
```

**Comment by sergekov** [comment]
I would vote up for this feature. It is extremely important to be able to select agent based on certain criteria in the cloud. Without it I am unable to limit a developer to certain environment such as dev/uat/staging. Therefore bad code can be executed and deployed in production.

**Comment by gtaylor** [comment]
@ozbillwang Only problem with defining this in the .drone.yaml is that I can sneak a change in there to run my build on a privileged agent.

If it's an agent env var, it gets a bit harder to sneak escalated builds in.

**Comment by bradrydzewski** [comment]
The solution we will (re) implement was briefly and successfully tested when we used stomp as the rpc protocol which accepts sql-like query expressions to filter messages. Some of the ground work has already been laid to enable this capability.

Why expressions? Because people have diverse needs including limiting by random fields such as ram, cpu count, gpu, installed software packages and more.

### Agent Changes

The agent must be updated to pass an expression string to the server. The grpc method definitions already have an expression field defined. When the agent requests the next build from the queue, we pass the filter object:

```
func (t *Client) Next(c context.Context, f Filter) (*Pipeline, error)
```

The agent should optionally send an expression:

```
Filter struct {
	Labels map[string]string `json:"labels"`
	Expr   string            `json:"expr"`
}
```

The expression should be provided to the agent using the DRONE_FILTER parameter.

### Server Changes

The server will receive the expression and can use this to filter which jobs it accepts from the queue. You can see where we currently filter by labels [here](https://github.com/drone/drone/blob/782688872970bc7dea53f167ce4181458892369e/server/rpc.go#L92:L99).

The filter function should be modified to use the filter expression or labels, but not both

```diff
+	var s, err
+	if filter.Epxr != "" {
+		s, err = selector.Parse(filter.Expr)
+		if err != nil {
+			return err
+		}
+	}


	fn := func(task *queue.Task) bool {
+		if s != nil {
+			ok, _ := s.Eval(task.Labels)
+			return ok
+		}
		for k, v := range filter.Labels {
			if task.Labels[k] != v {
				return false
			}
		}
		return true
	}
	task, err := s.queue.Poll(c, fn)
```

Note that the we provide key value pairs (labels) which represents the fields in the expression. The key is the field name. This is important (for later)

```
ok, _ := s.Eval(task.Labels)
```

The expression parse and evaluator comes from the very well tested library that we created for our stomp broker:
https://github.com/drone/mq/tree/master/stomp/selector

You can see a full list of supported syntax here:
https://github.com/drone/mq/blob/master/stomp/selector/selector_test.go

# Available Fields

Finally we need to ensure that we expose fields (labels) for the query selector to evaluate. Right now we only pass the platform so that servers can request architecture specific builds. You can see this logic [here](https://github.com/drone/drone/blob/782688872970bc7dea53f167ce4181458892369e/server/hook.go#L324:L328)

```
task.Labels = map[string]string{}
task.Labels["platform"] = item.Platform
```

This would allow the following expression:

```
platform == 'linux/amd64'
```

If we want to use other fields, we need to add them:

```diff
task.Labels = map[string]string{}
task.Labels["platform"] = item.Platform
+task.Labels["repo"]  = item.Repo.FullName
+task.Labels["environment"]  = item.Build.Environment # e.g. production
```

**Comment by gtaylor** [comment]
Just had a thought, wanted to clarify (with pseudo expressions that are probably not right).

Agent 1: DRONE_FILTER="repo-name == reddit/some-prefix-*"
Agent 2: <no drone repo filter>

How can we make sure that builds filtered on Agent 1 never end up on Agent 2? Would we do something like:

Agent 1: DRONE_FILTER="repo-name == reddit/some-prefix-*"
Agent 2: DRONE_FILTER="repo-name != reddit/some-prefix-*"

**Comment by bradrydzewski** [comment]
@gtaylor yes, exactly

although in this case you could use the glob keyword and its negative:

```sql
repo GLOB 'reddit/some-prefix-*'
repo NOT GLOB 'reddit/some-prefix-*'
```

you can see a bunch of syntax examples here:
https://github.com/drone/expr/blob/master/selector_test.go

**Comment by gtaylor** [comment]
Ah, excellent. This does more than I was initially thinking. Thanks!

**Comment by bradrydzewski** [comment]
ok I published an update to https://github.com/drone/expr so that it is easier to work with strings

Example creates the selector:

```
	var s, err
	if filter.Epxr != "" {
		s, err = expr.ParseString(filter.Expr)
		if err != nil {
			return err
		}
	}

```

Example converts the labels to a columnar row:

```
		if s != nil {
			ok, _ := s.Eval(expr.NewRow(task.Labels))
			return ok
		}
```

And that should really be it.

**Comment by bradrydzewski** [comment]
implemented by #2200
